### ISSUE TITLE
Efficient solver initialization by cloning

### ISSUE THREAD

**Issue by fpoli** [viperproject/silicon#731]
@utaal mentioned me that the Z3 API has a way to *clone* a solver. The method seems to be [Solver::translate](https://z3prover.github.io/api/html/classcom_1_1microsoft_1_1z3_1_1_solver.html#a2f4086956a17bb24964a5b0d53ff65f6). For the Silicon parallelization it might be more efficient to clone a solver when branching, instead of initializing a new solver from scratch (loading the prelude and all the assumptions coming from the path verified until that point). I'm not sure how easy/hard this is in the current architecture of Silicon.

**Comment by marcoeilers** [comment]
I had looked at this and thought it didn't do what we need in a concurrent setting, but I'll have another look.

**Comment by marcoeilers** [comment]
The problem is basically this:
We'd need to clone the solver before executing the then or else branch. However, at that point we don't know if we're going to execute the else branch in parallel (that depends on the availability of a free worker), so we'd likely clone many solvers that then never get used, which would cause some amount of overhead. 

So we'll have to just try that out to see if it's worth it. I'll put it on my to do list.

**Comment by jcp19** [comment]
If the overhead is actually high, we could try to use some sort of copy-on-write mechanism

**Comment by fpoli** [comment]
> We'd need to clone the solver before executing the then or else branch.

If the decision to parallelize happens after the branching point, it might be that cloning the solver and popping a few states (to go back to the state before the branch) is still faster than re-loading a solver from scratch.

**Comment by marcoeilers** [comment]
> > We'd need to clone the solver before executing the then or else branch.
> 
> If the decision to parallelize happens after the branching point, it might be that cloning the solver and popping a few states (to go back to the state before the branch) is still faster than re-loading a solver from scratch.

If I understand you correctly, you're saying we could clone the solver later, while the then-branch is potentially already being verified (and then pop some asserts to remove whatever we added while verifying the then-branch)? That would mean accessing the solver from two threads at the same time; I'm pretty sure we'd get segfaults left and right.

**Comment by fpoli** [comment]
> If I understand you correctly, you're saying we could clone the solver later, while the then-branch is potentially already being verified (and then pop some asserts to remove whatever we added while verifying the then-branch)? That would mean accessing the solver from two threads at the same time; I'm pretty sure we'd get segfaults left and right.

No, it was some confusion on my side regarding how Silicon internally works. I wanted to suggest that the solver can be cloned when queueing a verification task, but I now understand that because of work-stealing when queueing it's not known wether the task will be executed in parallel or not.

**Comment by marcoeilers** [comment]
I tried it.

Apparently, cloning solvers is not supported when using push/pop (I get the error ``translation of contexts is only supported at base level``). See https://github.com/Z3Prover/z3/issues/556.
